File: sync-client/packages/core/src/cache/prefetch-engine.ts
import { DEFAULT_CONCURRENCY, FILE_DOWNLOAD_TIMEOUT_MS } from '../constants.js';
import type { APIClient } from '../api-client.js';
import type { CacheManager } from './cache-manager.js';
import type { MetadataCache } from './metadata-cache.js';
import type { FileInfo } from '../types.js';
import type { PrefetchResult } from './types.js';
import type { LogCallback, ProgressCallback } from '../progress.js';

export interface PrefetchOptions {
  /** Playlist keys to prefetch. If empty, uses all pinned playlists. */
  playlists: string[];
  /** Output profile name for server-tagged downloads. */
  profile?: string;
  /** Number of parallel downloads. */
  concurrency?: number;
  /** Maximum cache size in bytes — runs eviction at end. 0 = unlimited. */
  maxCacheBytes?: number;
  /** Set of pinned playlist keys — used for eviction priority (unpinned evicted first). */
  pinnedPlaylists?: Set<string>;
  /** AbortSignal for cancellation. */
  signal?: AbortSignal;
  /** Progress callback. */
  onProgress?: ProgressCallback;
  /** Log callback. */
  onLog?: LogCallback;
  /** Optional metadata cache for ETag-based conditional requests. */
  metadataCache?: MetadataCache;
}

/**
 * Downloads pinned playlist files to the local cache (not to a destination).
 * Uses the same worker-pool concurrency pattern as SyncEngine.
 */
export class PrefetchEngine {
  private readonly client: APIClient;
  private readonly cacheManager: CacheManager;

  constructor(client: APIClient, cacheManager: CacheManager) {
    this.client = client;
    this.cacheManager = cacheManager;
  }

  async prefetch(options: PrefetchOptions): Promise<PrefetchResult> {
    const startTime = Date.now();
    const concurrency = options.concurrency ?? DEFAULT_CONCURRENCY;
    const log = options.onLog ?? (() => {});
    const onProgress = options.onProgress ?? (() => {});

    // Prune stale entries before starting
    const pruned = this.cacheManager.pruneStaleEntries();
    if (pruned > 0) {
      log('info', `Pruned ${pruned} stale cache entries`);
    }

    // Discover files for all playlists
    interface PlaylistFiles {
      key: string;
      files: FileInfo[];
    }
    const playlistFileList: PlaylistFiles[] = [];
    let grandTotal = 0;

    for (const key of options.playlists) {
      if (options.signal?.aborted) break;
      try {
        const response = await this.client.getFiles(key, false, options.profile, options.metadataCache);
        playlistFileList.push({ key, files: response.files });
        grandTotal += response.files.length;
      } catch (err) {
        log('warn', `Skipping playlist "${key}": ${err}`);
      }
    }

    const hasLimit = options.maxCacheBytes !== undefined && options.maxCacheBytes > 0;
    const maxCacheBytes = options.maxCacheBytes ?? 0;

    // Pre-filter: determine which files would be evicted anyway due to capacity
    const capacityExcluded = new Set<string>();

    if (hasLimit) {
      // Deduplicate server files by UUID (a file may appear in multiple playlists)
      const seenUuids = new Set<string>();
      const allFiles: { uuid: string; size: number; createdAt: number }[] = [];
      for (const { files } of playlistFileList) {
        for (const file of files) {
          if (!seenUuids.has(file.uuid)) {
            seenUuids.add(file.uuid);
            allFiles.push({
              uuid: file.uuid,
              size: file.size,
              createdAt: file.created_at ?? 0,
            });
          }
        }
      }

      // Sort newest first (inverse of eviction order)
      allFiles.sort((a, b) => b.createdAt - a.createdAt);

      // Files whose cumulative size exceeds the limit would be evicted anyway
      let cumulative = 0;
      for (const f of allFiles) {
        cumulative += f.size;
        if (cumulative > maxCacheBytes) {
          capacityExcluded.add(f.uuid);
        }
      }
    }

    // Filter out already-cached files and capacity-excluded files
    const toDownload: { key: string; file: FileInfo }[] = [];
    let totalSkipped = 0;
    let capacityCapped = 0;

    for (const { key, files } of playlistFileList) {
      for (const file of files) {
        if (capacityExcluded.has(file.uuid)) {
          capacityCapped++;
          continue;
        }
        const cached = this.cacheManager.isCached(file.uuid);
        const stale = cached && this.cacheManager.isStale(file.uuid, file.updated_at);
        if (cached && !stale) {
          totalSkipped++;
        } else {
          toDownload.push({ key, file });
        }
      }
    }

    if (hasLimit) {
      const currentSize = this.cacheManager.getTotalSize();
      const available = maxCacheBytes - currentSize;
      log('info', `Cache: ${formatBytes(currentSize)} used / ${formatBytes(maxCacheBytes)} limit (${formatBytes(Math.max(0, available))} available)`);
    }

    log('info', `Prefetch: ${toDownload.length} to download, ${totalSkipped} already cached` + (capacityCapped > 0 ? `, ${capacityCapped} exceeded cache capacity` : ''));

    onProgress({
      phase: 'syncing',
      processed: totalSkipped + capacityCapped,
      total: grandTotal,
      copied: 0,
      skipped: totalSkipped,
      failed: 0,
    });

    // Download with concurrency limit — stop when cache is full
    let downloaded = 0;
    let failed = 0;
    let processed = totalSkipped + capacityCapped;
    let aborted = false;
    let capacityReached = false;
    let index = 0;
    const downloadedUuids = new Set<string>();

    const worker = async () => {
      while (index < toDownload.length) {
        if (options.signal?.aborted) {
          aborted = true;
          return;
        }
        if (capacityReached) {
          return;
        }
        const item = toDownload[index++]!;
        const success = await this.downloadToCache(item.key, item.file, options.profile, options.signal, log);
        processed++;
        if (success) {
          downloaded++;
          downloadedUuids.add(item.file.uuid);
          // Check if cache is now full — try eviction to make room
          if (hasLimit && this.cacheManager.getTotalSize() >= maxCacheBytes) {
            // Step 1: evict unpinned files first (cheapest to lose)
            if (options.pinnedPlaylists && options.pinnedPlaylists.size > 0) {
              let overage = this.cacheManager.getTotalSize() - maxCacheBytes;
              const freed = this.cacheManager.evictUnpinnedBytes(overage, options.pinnedPlaylists);
              if (freed > 0) {
                log('info', `Evicted ${formatBytes(freed)} of unpinned cache to make room`);
              }
            }
            // Step 2: if still over, evict oldest files (skip this session's downloads)
            if (this.cacheManager.getTotalSize() >= maxCacheBytes) {
              const overage = this.cacheManager.getTotalSize() - maxCacheBytes;
              const freed = this.cacheManager.evictOldestBytes(overage, downloadedUuids);
              if (freed > 0) {
                log('info', `Evicted ${formatBytes(freed)} of oldest cache to make room`);
              }
            }
            // After all eviction attempts, check if still over limit
            if (this.cacheManager.getTotalSize() >= maxCacheBytes) {
              capacityReached = true;
              const remaining = toDownload.length - index;
              if (remaining > 0) {
                capacityCapped += remaining;
                processed += remaining;
                log('info', `Cache limit reached — skipping ${remaining} remaining files`);
              }
            }
          }
        } else {
          failed++;
        }
        onProgress({
          phase: 'syncing',
          playlist: item.key,
          file: item.file.display_filename || item.file.filename,
          processed,
          total: grandTotal,
          copied: downloaded,
          skipped: totalSkipped,
          failed,
        });
      }
    };

    const workers = Array.from(
      { length: Math.min(concurrency, toDownload.length) },
      () => worker(),
    );
    await Promise.all(workers);

    // Evict if over limit (skip if all downloads were capacity-capped — nothing to evict)
    let evictedBytes = 0;
    if (hasLimit && !capacityReached) {
      evictedBytes = this.cacheManager.evictToLimit(maxCacheBytes, options.pinnedPlaylists);
      if (evictedBytes > 0) {
        log('info', `Evicted ${formatBytes(evictedBytes)} to stay within cache limit`);
      }
    }

    onProgress({
      phase: aborted ? 'aborted' : 'complete',
      processed,
      total: grandTotal,
      copied: downloaded,
      skipped: totalSkipped,
      failed,
    });

    return {
      downloaded,
      skipped: totalSkipped,
      failed,
      capacityCapped,
      aborted,
      durationMs: Date.now() - startTime,
    };
  }

  private async downloadToCache(
    playlistKey: string,
    file: FileInfo,
    profile: string | undefined,
    signal: AbortSignal | undefined,
    log: LogCallback,
  ): Promise<boolean> {
    try {
      const timeoutSignal = AbortSignal.timeout(FILE_DOWNLOAD_TIMEOUT_MS);
      const combinedSignal = signal
        ? AbortSignal.any([signal, timeoutSignal])
        : timeoutSignal;
      const { body } = await this.client.downloadFile(playlistKey, file.filename, profile, combinedSignal);
      await this.cacheManager.storeStream(file, playlistKey, body, file.created_at, file.updated_at);
      return true;
    } catch (err) {
      if (signal?.aborted) return false;
      const isTimeout = err instanceof DOMException && err.name === 'TimeoutError';
      if (isTimeout) {
        log('warn', `Prefetch timed out for ${playlistKey}/${file.filename}`);
      } else {
        log('error', `Prefetch failed for ${playlistKey}/${file.filename}: ${err}`);
      }
      return false;
    }
  }
}

const BYTES_PER_KB = 1024;
const BYTES_PER_MB = 1024 * 1024;
const BYTES_PER_GB = 1024 * 1024 * 1024;

function formatBytes(bytes: number): string {
  if (bytes >= BYTES_PER_GB) return `${(bytes / BYTES_PER_GB).toFixed(1)} GB`;
  if (bytes >= BYTES_PER_MB) return `${(bytes / BYTES_PER_MB).toFixed(1)} MB`;
  if (bytes >= BYTES_PER_KB) return `${(bytes / BYTES_PER_KB).toFixed(1)} KB`;
  return `${bytes} B`;
}

File: clients/ios/MusicPorter/MusicPorter/Services/Cache/PrefetchEngine.swift
import Foundation
import os

/// Downloads playlist files to the local cache (not to a destination).
///
/// Port of sync client's prefetch-engine.ts. Uses Swift structured concurrency
/// (TaskGroup) for concurrent downloads and Task cancellation for abort.
actor PrefetchEngine {
    private let apiClient: APIClient
    private let cacheManager: AudioCacheManager
    private let logger = Logger(subsystem: "com.musicporter", category: "PrefetchEngine")

    init(apiClient: APIClient, cacheManager: AudioCacheManager) {
        self.apiClient = apiClient
        self.cacheManager = cacheManager
    }

    /// Prefetch files for the given playlists into the local cache.
    func prefetch(
        options: PrefetchOptions,
        metadataCache: MetadataCache,
        onProgress: (@Sendable (PrefetchProgress) -> Void)? = nil
    ) async -> PrefetchResult {
        let startTime = ContinuousClock.now

        // Prune stale entries before starting
        let pruned = await cacheManager.pruneStaleEntries()
        if pruned > 0 {
            logger.info("Pruned \(pruned) stale cache entries")
        }

        // Discover files for all playlists
        var playlistFileList: [(key: String, files: [Track])] = []
        var grandTotal = 0

        for key in options.playlists {
            if Task.isCancelled { break }
            do {
                let response = try await apiClient.getFilesWithETag(
                    playlist: key,
                    profile: options.profile,
                    metadataCache: metadataCache
                )
                playlistFileList.append((key: key, files: response.files))
                grandTotal += response.files.count
            } catch {
                logger.warning("Skipping playlist \"\(key)\": \(error)")
            }
        }

        let hasLimit = options.maxCacheBytes > 0

        // Pre-filter: determine which files would be evicted anyway due to capacity
        var capacityExcluded = Set<String>()

        if hasLimit {
            // Deduplicate server files by UUID
            var seenUuids = Set<String>()
            var allFiles: [(uuid: String, size: Int, createdAt: Double)] = []
            for (_, files) in playlistFileList {
                for file in files {
                    guard let uuid = file.uuid, !seenUuids.contains(uuid) else { continue }
                    seenUuids.insert(uuid)
                    allFiles.append((uuid: uuid, size: file.size, createdAt: file.createdAt ?? 0))
                }
            }

            // Sort newest first (inverse of eviction order)
            allFiles.sort { $0.createdAt > $1.createdAt }

            // Files whose cumulative size exceeds the limit would be evicted anyway
            var cumulative: Int64 = 0
            for f in allFiles {
                cumulative += Int64(f.size)
                if cumulative > options.maxCacheBytes {
                    capacityExcluded.insert(f.uuid)
                }
            }
        }

        // Filter out already-cached files and capacity-excluded files
        var toDownload: [(key: String, file: Track)] = []
        var totalSkipped = 0
        var capacityCapped = 0

        for (key, files) in playlistFileList {
            for file in files {
                guard let uuid = file.uuid else { continue }
                if capacityExcluded.contains(uuid) {
                    capacityCapped += 1
                    continue
                }
                let cached = await cacheManager.isCached(uuid)
                var stale = false
                if cached != nil {
                    stale = await cacheManager.isStale(uuid: uuid, serverUpdatedAt: file.updatedAt)
                }
                if cached != nil && !stale {
                    totalSkipped += 1
                } else {
                    toDownload.append((key: key, file: file))
                }
            }
        }

        if hasLimit {
            let currentSize = await cacheManager.getTotalSize()
            let available = max(0, options.maxCacheBytes - currentSize)
            logger.info("Cache: \(CacheUtils.formatBytes(currentSize)) used / \(CacheUtils.formatBytes(options.maxCacheBytes)) limit (\(CacheUtils.formatBytes(available)) available)")
        }

        logger.info("Prefetch: \(toDownload.count) to download, \(totalSkipped) already cached\(capacityCapped > 0 ? ", \(capacityCapped) exceeded cache capacity" : "")")

        onProgress?(PrefetchProgress(
            phase: "syncing",
            playlist: nil,
            file: nil,
            processed: totalSkipped + capacityCapped,
            total: grandTotal,
            downloaded: 0,
            skipped: totalSkipped,
            failed: 0
        ))

        // Download with concurrency limit
        let downloaded = LockedCounter()
        let failed = LockedCounter()
        let processedCounter = LockedCounter(initial: totalSkipped + capacityCapped)
        var aborted = false
        var capacityReached = false
        let downloadedUuids = LockedSet()
        let pinnedPlaylists = options.pinnedPlaylists

        // Use a simple index-based approach with TaskGroup
        let downloadIndex = LockedCounter()
        let concurrency = min(options.concurrency, toDownload.count)

        if concurrency > 0 {
            await withTaskGroup(of: Void.self) { group in
                for _ in 0..<concurrency {
                    group.addTask { [self] in
                        while true {
                            if Task.isCancelled { return }
                            if capacityReached { return }

                            let idx = downloadIndex.increment() - 1
                            guard idx < toDownload.count else { return }
                            let item = toDownload[idx]

                            let success = await self.downloadToCache(
                                playlistKey: item.key,
                                file: item.file,
                                profile: options.profile
                            )

                            let currentProcessed = processedCounter.increment()
                            if success {
                                let currentDownloaded = downloaded.increment()
                                if let uuid = item.file.uuid {
                                    downloadedUuids.insert(uuid)
                                }

                                // Check if cache is now full
                                if hasLimit {
                                    let totalSize = await self.cacheManager.getTotalSize()
                                    if totalSize >= options.maxCacheBytes {
                                        // Step 1: evict unpinned files first
                                        if !pinnedPlaylists.isEmpty {
                                            let overage = totalSize - options.maxCacheBytes
                                            let freed = await self.cacheManager.evictUnpinnedBytes(
                                                targetBytes: overage,
                                                pinnedPlaylists: pinnedPlaylists
                                            )
                                            if freed > 0 {
                                                self.logger.info("Evicted \(CacheUtils.formatBytes(freed)) of unpinned cache to make room")
                                            }
                                        }
                                        // Step 2: if still over, evict oldest (skip this session's downloads)
                                        let sizeAfterUnpinned = await self.cacheManager.getTotalSize()
                                        if sizeAfterUnpinned >= options.maxCacheBytes {
                                            let overage = sizeAfterUnpinned - options.maxCacheBytes
                                            let freed = await self.cacheManager.evictOldestBytes(
                                                targetBytes: overage,
                                                protectedUuids: downloadedUuids.values()
                                            )
                                            if freed > 0 {
                                                self.logger.info("Evicted \(CacheUtils.formatBytes(freed)) of oldest cache to make room")
                                            }
                                        }
                                        // Check if still over limit after all eviction
                                        let finalSize = await self.cacheManager.getTotalSize()
                                        if finalSize >= options.maxCacheBytes {
                                            capacityReached = true
                                            let remaining = toDownload.count - (idx + 1)
                                            if remaining > 0 {
                                                capacityCapped += remaining
                                                _ = processedCounter.add(remaining)
                                                self.logger.info("Cache limit reached — skipping \(remaining) remaining files")
                                            }
                                        }
                                    }
                                }

                                onProgress?(PrefetchProgress(
                                    phase: "syncing",
                                    playlist: item.key,
                                    file: item.file.displayFilename ?? item.file.filename,
                                    processed: currentProcessed,
                                    total: grandTotal,
                                    downloaded: currentDownloaded,
                                    skipped: totalSkipped,
                                    failed: failed.value
                                ))
                            } else {
                                let currentFailed = failed.increment()
                                onProgress?(PrefetchProgress(
                                    phase: "syncing",
                                    playlist: item.key,
                                    file: item.file.displayFilename ?? item.file.filename,
                                    processed: currentProcessed,
                                    total: grandTotal,
                                    downloaded: downloaded.value,
                                    skipped: totalSkipped,
                                    failed: currentFailed
                                ))
                            }
                        }
                    }
                }
            }
        }

        aborted = Task.isCancelled

        // Post-download eviction
        if hasLimit && !capacityReached {
            let evicted = await cacheManager.evictToLimit(
                maxBytes: options.maxCacheBytes,
                pinnedPlaylists: pinnedPlaylists.isEmpty ? nil : pinnedPlaylists
            )
            if evicted > 0 {
                logger.info("Evicted \(CacheUtils.formatBytes(evicted)) to stay within cache limit")
            }
        }

        let finalPhase = aborted ? "aborted" : "complete"
        onProgress?(PrefetchProgress(
            phase: finalPhase,
            playlist: nil,
            file: nil,
            processed: processedCounter.value,
            total: grandTotal,
            downloaded: downloaded.value,
            skipped: totalSkipped,
            failed: failed.value
        ))

        let elapsed = ContinuousClock.now - startTime
        let durationMs = Int(elapsed.components.seconds * 1000
            + elapsed.components.attoseconds / 1_000_000_000_000_000)

        return PrefetchResult(
            downloaded: downloaded.value,
            skipped: totalSkipped,
            failed: failed.value,
            capacityCapped: capacityCapped,
            aborted: aborted,
            durationMs: durationMs
        )
    }

    // MARK: - Internal

    private func downloadToCache(
        playlistKey: String,
        file: Track,
        profile: String?
    ) async -> Bool {
        do {
            let data = try await apiClient.downloadFileData(
                playlist: playlistKey,
                filename: file.filename,
                profile: profile
            )
            await cacheManager.storeData(
                data,
                file: file,
                playlistKey: playlistKey,
                serverCreatedAt: file.createdAt,
                serverUpdatedAt: file.updatedAt
            )
            return true
        } catch {
            if Task.isCancelled { return false }
            logger.error("Prefetch failed for \(playlistKey)/\(file.filename): \(error)")
            return false
        }
    }
}

// MARK: - Thread-safe Counters

/// Simple thread-safe counter using os_unfair_lock for use in TaskGroup.
private final class LockedCounter: @unchecked Sendable {
    private var _value: Int
    private let lock = NSLock()

    init(initial: Int = 0) { _value = initial }

    var value: Int {
        lock.lock()
        defer { lock.unlock() }
        return _value
    }

    /// Increment and return the new value.
    @discardableResult
    func increment() -> Int {
        lock.lock()
        defer { lock.unlock() }
        _value += 1
        return _value
    }

    /// Add amount and return the new value.
    @discardableResult
    func add(_ amount: Int) -> Int {
        lock.lock()
        defer { lock.unlock() }
        _value += amount
        return _value
    }
}

/// Thread-safe set of strings for tracking downloaded UUIDs.
private final class LockedSet: @unchecked Sendable {
    private var _values = Set<String>()
    private let lock = NSLock()

    func insert(_ value: String) {
        lock.lock()
        defer { lock.unlock() }
        _values.insert(value)
    }

    func values() -> Set<String> {
        lock.lock()
        defer { lock.unlock() }
        return _values
    }
}
